### PR TITLE
cumall and cumany handle NA values correctly

### PR DESCRIPTION
**Pull request by s-fleck**
`cumall()` and `cumany()` do not behave as outlined in (#408). `cumall(TRUE, NA, FALSE)` returns `TRUE, NA, NA` instead of `TRUE, NA, FALSE`. `cumany()` behaviour is similarly broken. The respective unit tests were ineffective.

I modified the c++ src file and added stricter unit tests; however, I don't do much work in RCPP so I hope I didn't to anything stupid in my attempt to fix this issue.

### CODE REVIEW

File: src/window.cpp [hunk 1]
```diff
@@ -13,23 +13,27 @@
...
+    out[0] = NA_LOGICAL;
+
   for (int i = 1; i < n; i++) {
     current = x[i];
-    if (current == NA_LOGICAL) break;
-    if (current == FALSE) {
-      std::fill(out.begin() + i, out.end(), FALSE);
+
+    if (current == TRUE)
+      out[i] = out[i - 1];
```
Comment (krlmlr) [inline src/window.cpp]: Can we replace the reference to `out[i - 1]` with a local variable that is suitably initialized and updated? This might allow getting rid of the special case for `i == 0` .

File: tests/testthat/test-window.R [hunk 2]
```diff
@@ -12,16 +12,30 @@ test_that("If n = length(x), returns all missing", {
   expect_equal(lag(1:2, 2), miss)
 })
 
-test_that("cumany handles NA (#408)", {
+test_that("cumany and cumall handle NA (#408)", {
```
Comment (krlmlr) [inline tests/testthat/test-window.R]: Can you please add a reference to this PR?

File: src/window.cpp (hunk 1 above)
Comment (s-fleck) [reply src/window.cpp]: Hmm I am not sure exactly how to do that gracefully, I usually don't code in c++. I am open to suggestions though :)

File: tests/testthat/test-window.R (hunk 2 above)
Comment (s-fleck) [reply tests/testthat/test-window.R]: done

File: src/window.cpp (hunk 1 above)
Comment (krlmlr) [reply src/window.cpp]: Something like:

```
int previous = TRUE;
for (...) {
  if (current)
    out[i] = previous;
  else if (current == NA_LOGICAL)
    previous = out[i] = NA_LOGICAL;
  else
    ........
```

Comment (s-fleck) [reply src/window.cpp]: done... also all nonzero values in C are `TRUE` (like `NA_LOGICAL`), so it must stay `if (current == TRUE)`

Comment (krlmlr) [reply src/window.cpp]: Yes, of course. Maybe:

```r
if (current == NA_LOGICAL) {
  previous = NA_LOGICAL;
  out[i] = NA_LOGICAL;
}
else if (current)
  out[i] = previous;
```

(turning the double assignment into two separate assignments to make the code clearer.

File: src/window.cpp [hunk 3]
```diff
@@ -14,22 +14,16 @@
 LogicalVector cumall(LogicalVector x) {
   int n = x.length();
   LogicalVector out(n, FALSE);
+  int current  = out[0] = x[0];
```
Comment (krlmlr) [inline src/window.cpp]: I think `int current = FALSE` would work, in particular it's necessary for `n == 0` .

File: src/window.cpp [hunk 4]
```diff
@@ -13,23 +13,21 @@
 // [[Rcpp::export]]
 LogicalVector cumall(LogicalVector x) {
   int n = x.length();
-  LogicalVector out(n, NA_LOGICAL);
+  LogicalVector out(n, FALSE);
+  int current  = out[0] = x[0];
```
Comment (krlmlr) [inline src/window.cpp]: Sorry: `int current = TRUE` .

Comment (s-fleck) [reply src/window.cpp]: Hmm wouldn't that break `cumall(FALSE)`?

I guess your right that the code is sketchy for `n = 0`, but it does return correctly (`logical()`)
I added the respective tests

Comment (krlmlr) [reply src/window.cpp]: Why would the results be different from `cumall(c(FALSE, NA, TRUE))` which is tested by your tests? Testing these "simple" corner cases is a very good thing, btw!

`out[0]` and `x[0]` are undefined for `n == 0` , tools like `valgrind` will detect this.

Comment (s-fleck) [reply src/window.cpp]: If initialize current and previous with `TRUE` woudn't that mean that the first element of `x` is never checked? That would break all vectors that start with `FALSE` or `NA`.

Why not just a `if (n == 0) return (out)` before initializing current?

Comment (s-fleck) [reply src/window.cpp]: Right, I forgot that `current` is re-assigned in the loop anyways.

**Comment by romainfrancois** [comment]
@s-fleck can you please rebase on the current master (which is free of #3761) so that travis can check the changes.

**Comment by romainfrancois** [comment]
Thanks. Looks good (no worries, still not completely comfortable with rebasing too)
 
I trust travis more than appveyor anyway. I'll have a look at the code soon.

File: src/window.cpp [hunk 5]
```diff
@@ -13,23 +13,22 @@
...
-  }
-  for (int i = 1; i < n; i++) {
+  for (int i = 0; i < n; i++) {
     current = x[i];
-    if (current == NA_LOGICAL) break;
-    if (current == FALSE) {
-      std::fill(out.begin() + i, out.end(), FALSE);
+    if (current == TRUE){
+      out[i] = previous;
+    } else if (current == NA_LOGICAL){
```
Comment (krlmlr) [inline src/window.cpp]: Maybe switch order so that we hedge against degenerated logicals? Not sure how relevant this is in practice.

```c++
if (current == NA_LOGICAL) {
} else if (current) {
} else {
}
```

Comment (s-fleck) [reply src/window.cpp]: Hmm what do you mean by degenerate logicals? logicals that are not `0`, `1` or `NA_LOGICAL`?. Currently they go through all the way to the `break`. `current == TRUE` should only evaluate to `TRUE` for `1` if I understand it correclty (and as opposed to if (current) ). If we want to check that, maybe we should make the last `else` an `else if (current == FALSE)` and add an `else { ...throw exception...} ` at the end?

Comment (krlmlr) [reply src/window.cpp]: Yes. I'm not sure how relevant such degenerate logicals are in practice. These could be returned by an R package that generates a data frame in native code and is not careful to use only `TRUE`, `FALSE` or `NA` .

Comment (krlmlr) [reply src/window.cpp]: Maybe:

```c++
if (current == NA_LOGICAL) {
} else if (current == FALSE) {
} else {
}
```

R itself isn't very careful with degenerate logicals, so it may be a non-issue, because native code that creates logical vectors will probably do the right thing:

``` r
logical_two <- Rcpp::cppFunction(
  "LogicalVector test() { LogicalVector ret(1); ret[0] = 2; return ret; }"
)

x <- logical_two()
as.integer(x)
#> [1] 2
sum(x)
#> [1] 2
```

> Created on 2018-09-06 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).
